substrate-airgapped/src/tx/generic_call.rs

```diff
@@ -55,8 +55,11 @@ where
 
 impl<C: Encode + Decode + Clone> Encode for GenericCall<C> {
 	fn encode(&self) -> Vec<u8> {
-		let mut bytes = self.call_index.to_vec();
-		bytes.extend(self.args.encode());
+		let encoded_args = self.args.encode();
+		let encoded_call_index = self.call_index.to_bytes();
+		let mut bytes = Vec::with_capacity(encoded_call_index.len() + encoded_args.len());
+		bytes.extend_from_slice(&encoded_call_index);
+		bytes.extend_from_slice(&encoded_args);
 
```

Comment:
Instead of using the [exact code @maciejhirsz prescribed](https://github.com/paritytech/substrate-airgapped/compare/mh-review#diff-c86dc94950b482469444ddd8b6e8068a0a28761909f3e33d1a492f52b8fee35dR64-R73), I opted to get the decoded values up front for 2 reasons:

1) While we know for now the `call_index` is a fixed size array of length 2, I think it is more future proof to just take the length from the encoded value (from what I read-on-the-internet `.len()` is O(1) for vecs and arrays)
1) `parity-scale-codec`'s `Encode::encoded_size` is O(n) time complexity (assuming the compiler doesn't optimize here), so I think it makes more sense just to get the `encoded` value first so we can then get its length in constant time

Reply:
Well, as long as the `to_bytes` signature says `[u8; 2]` you're not actually future proofing anything I think? 
`len()` for an array should be a no-op (i.e. resolved at compile time) but I'm not positive that it is.

Either way, what you've done here is fine too.

Reply:
`arrays` is using [slice::len](https://doc.rust-lang.org/std/primitive.slice.html#method.len) which is const fn, thus it's cheap

Reply:
`.len()` on an array should basically compile to a constant, so that's cool.

`Encode::encoded_size` is not quite `O(n)`, the complexity depends on what it is that you are encoding, for a single struct it's going to be `O(1)`, f.e. `Encode::encode` uses `size_hint` for capacity from what I see in code, which is not the same as `encoded_size`, so for encoding to `Vec` it probably has to do reallocations while encoding (which is worse than `O(n)`). AFAIK `bincode` (very similar to SCALE) has been pre-allocating exact capacity in the `Vec` for a long while now, and doing so was a pretty big performance boost when it was merged. Either way, best way to know is to benchmark it :).

Reply:
I went to go change this to Maciej's original suggestion with `decoded_size`, but after awhile of the compiler complaining I realized that its only available in `parity-scale-codec` 2.0.0. So I tried bumping to 2.0.0, but then I realized that sp-* deps all use `parity-scale-codec` 1.3.6 or lower and there are some conflicts with the way `Encode` and `Decode` are defined on some of the imports we use.

Reply:
substrate master has upgraded to scale 2.0 now so you could try again.

To add to what @maciejhirsz said, I think the plan is to optimize the `encoded_size` further and what we see today in code is the ground work for that.

Reply:
@dvdplm should we be tracking master in airgapped? Seems like it would be impossible to audit